PYME/recipes/kinetics_recipes.py

```diff
@@ -0,0 +1,428 @@
+        pylab.figtext(.6, .8 - .05, 'tau = %3.4f' % (res_on[0][1]), size=18, color=colours[1])
+        pylab.show()
+
+
+        pylab.figure()
+
+        pylab.xlim(right=np.round(self.maxOffTimeSecondsToFit/cyctime).astype('int'))
+        pylab.bar(bins_off * cyctime, n_off, width=np.diff(bins_off)[0] * cyctime, alpha=0.3, fc=colours[2])
+        pylab.plot(np.linspace(bins_off[0], max(bins_off), max(bins_off) * 10) * cyctime,
+                   eimod(res_off[0], np.linspace(1, max(bins_off), max(bins_off) * 10) * cyctime),
+                   colours[3], lw=3, alpha=.8)
+        pylab.ylabel('Events')
+        pylab.xlabel('Event Duration [s]')
+        pylab.ylim((.9, pylab.ylim()[1]))
+        pylab.title('Blink dark times')
+
+        pylab.figtext(.6, .8 - .05, 'tau = %3.4f' % (res_off[0][1]), size=18, color=colours[1])
+
+        pylab.show()
+
+        powers = np.array([res_on[0][1], res_off[0][1]])
+        times = np.array([cyctime, (max(ds[oncol]).astype(float)+1)*cyctime, (max(ds[offcol]).astype(float)+1)*cyctime])
+
+        output_list = [powers, times, n_on, n_off]
+
+        namespace[self.outputEmphistData] = output_list
+
+@register_module('save_EmpHistJSON')
+class save_EmpHistJson(ModuleBase):
+    """
+    Creates empirical histogram files for use with the simulator, based on the outputs from the fitting module above.
+    """
+    inputName = Input('emphist_data')
+    outputdir = CStr('C:\\Users\\bdr25\\Desktop\\data\\JSON files')
+    outputName = Output('emphist_file')
+
+    def execute(self, namespace):
+        import io
+        import json
+        import time
+
+        powers, times, n_on, n_off = namespace[self.inputName]
+
+        new_on_nums = np.zeros(len(n_on)+1)
+        new_off_nums = np.zeros(len(n_off)+1)
+        new_on_nums[1:] = n_on
+        new_off_nums[1:] = n_off
+
+        json_prepped_on = np.flip((new_on_nums.reshape((len(new_on_nums), 1))).astype('int'), axis=1).tolist()
+        json_prepped_off = np.flip((new_off_nums.reshape((len(new_off_nums), 1))).astype('int'), axis=1).tolist()
+
+        dk = {}
+        dk['plog'] = [0, 0]
+        dk['tlog'] = [0, 0]
+        dk['pmin'] = [min(powers), min(powers)]
+        dk['pmax'] = [max(powers), max(powers)]
+        dk['tmin'] = [times[0], times[0]]
+        dk['tmax'] = [times[1], times[2]]
+        dk['on'] = json_prepped_on
+        dk['off'] = json_prepped_off
+
+        output_dir = self.outputdir
+
+        try:
+            to_unicode = unicode
+        except NameError:
+            to_unicode = str
+
+        timestr = time.strftime("%Y%m%d-%H%M%S")
+
+        with io.open(output_dir + '\\' + 'test_emphist' + timestr + '.json', 'w', encoding='utf8') as outfile:
+            str_ = '{"dk": ' + json.dumps(dk) + '}'
+            outfile.write(to_unicode(str_))
+
+        print('Json has been created and saved to disk')
+
+
+@register_module('HDBSCANClustering')
+class HDBSCANClustering(ModuleBase):
```

Comment:
This should probably be right next to the `localisations.DBSCANClustering` module rather than another file. Perhaps even all clustering modules should be in a new file, recipes/clustering, such that they appear next to each other in the UI without naming them backwards, e.g. `localisations.ClusteringDBSCAN` and `localisations.ClusteringHDBSCAN`. If I recall, @David-Baddeley will prefer not the backwards naming, so maybe this is a good time to move them to a separate file (or maybe no one cares if they aren't next to each other in the UI).